modules/v2x/v2x_proxy/proto_adapter/proto_adapter.cc

```diff
@@ -0,0 +1,230 @@
+/******************************************************************************
+ * Copyright 2020 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/v2x/v2x_proxy/proto_adapter/proto_adapter.h"
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+#include "cyber/cyber.h"
+
+namespace v2x {
+namespace proxy {
+
+OSLightype AdapterLightTypeObu2Sys(int32_t type) {
+  switch (type) {
+    case 2:
+      return OSLightype::SingleTrafficLight_Type_LEFT;
+    case 3:
+      return OSLightype::SingleTrafficLight_Type_RIGHT;
+    case 4:
+      return OSLightype::SingleTrafficLight_Type_U_TURN;
+    case 1:
+    default:
+      return OSLightype::SingleTrafficLight_Type_STRAIGHT;
+  }
+}
+
+bool AdapterLightObu2Sys(const OBULight &obu_light,
+                         std::shared_ptr<OSLight> *os_light_ptr) {
+#if 0
+  if (nullptr == os_light_ptr) {
+    return false;
+  }
+  auto res = std::make_shared<OSLight>();
+  if (nullptr == res) {
+    return false;
+  }
+  if (!obu_light.has_header()) {
+    return false;
+  }
+  if (obu_light.header().has_timestamp_sec()) {
+    res->mutable_header()->set_timestamp_sec(
+        obu_light.header().timestamp_sec());
+  }
+  if (obu_light.header().has_module_name()) {
+    res->mutable_header()->set_module_name(obu_light.header().module_name());
+  }
+  if (obu_light.road_trafficlight_size() < 1) {
+    return false;
+  }
+  //// FOR-EACH ROAD
+  bool flag_has_data = false;
+  for (int idx_road = 0; idx_road < obu_light.road_trafficlight_size();
+       idx_road++) {
+    const auto &obu_road_light1 = obu_light.road_trafficlight(idx_road);
```

Comment:
traffic_light
And why a `light1` here, since there is no `light0` or `light2`.

Reply:
Due to historical reasons, the proto definition was changed from _optional_ to _repeated_ , and the code changed from "processing the first one only" to "processing one by one" . So here emphasize "one" instead of "that"